### PR TITLE
Use GetForUpdate on unique indexes

### PR DESCRIPTION
**Pull request by graetzer**
### Scope & Purpose

Use GetForUpdate when inserting unique index entries. This performs the unique-check, locking and write-write conflict check in one call to rocksdb. Then we set `assume_tracked=true`, to avoid the additional read on disk that would usually occur with the normal `Transaction::Put`.

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behaviour change can only be verified via automatic tests

### Testing & Verification

This change is already covered by existing tests, such as all existing integration tests

http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/6399/

### Documentation

- [x] Added a *Changelog Entry*

### CODE REVIEW

File: arangod/RocksDBEngine/RocksDBVPackIndex.cpp [hunk 1]
```diff
@@ -702,18 +702,20 @@ Result RocksDBVPackIndex::insert(transaction::Methods& trx, RocksDBMethods* mthd
   RocksDBValue value = _unique ? RocksDBValue::UniqueVPackIndexValue(documentId)
                                : RocksDBValue::VPackIndexValue();
 
-  size_t const count = elements.size();
-  rocksdb::PinnableSlice existing;
-  for (size_t i = 0; i < count; ++i) {
+#warning move loop up
```
Comment (danielhlarkin) [inline arangod/RocksDBEngine/RocksDBVPackIndex.cpp]: Remove? Or explain.

File: arangod/RocksDBEngine/RocksDBVPackIndex.cpp [hunk 2]
```diff
@@ -702,18 +702,20 @@ Result RocksDBVPackIndex::insert(transaction::Methods& trx, RocksDBMethods* mthd
...
+  const bool unique = _unique;
```
Comment (danielhlarkin) [inline arangod/RocksDBEngine/RocksDBVPackIndex.cpp]: Why are we creating a copy here?

**Comment by jsteemann** [comment]
Just did a very incomplete benchmark run comparing devel and this branch:
*devel*
```
arangod> db._drop("abc"); db._create("abc"); time = require("internal").time; s = time(); for (i = 0; i < 1000 * 1000; ++i) db.abc.insert({ _key: "test" + i, value: i }); time() - s;
32.33614945411682
arangod> db._drop("abc"); db._create("abc"); time = require("internal").time; s = time(); for (i = 0; i < 1000 * 1000; ++i) db.abc.insert({ _key: "test" + i, value: i }); time() - s;
34.20624136924744
```
*this branch*:
```
arangod> db._drop("abc"); db._create("abc"); time = require("internal").time; s = time(); for (i = 0; i < 1000 * 1000; ++i) db.abc.insert({ _key: "test" + i, value: i }); time() - s;
32.889970779418945
arangod> db._drop("abc"); db._create("abc"); time = require("internal").time; s = time(); for (i = 0; i < 1000 * 1000; ++i) db.abc.insert({ _key: "test" + i, value: i }); time() - s;
38.492968797683716
```
It may look completely different when there are more (unique) indexes present, but with just the primary index and very small documents there doesn't seem to be a substantial speedup.

**Comment by graetzer** [comment]
It might be the case that the Get itself is actually not that expensive. Anyhow I will benchmark this on the weekend

**Comment by jsteemann** [comment]
@graetzer : sounds good. It may actually be beneficial for larger datasets or when there are more indexes present. Let's see.

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/6402/

**Comment by graetzer** [comment]
I also ran some benchmarks, in essence the insert performance was about the same. I think we should still merge this, because in theory now we can fail earlier if there is a write-write conflict

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/4862/

**Pull request merged**